### PR TITLE
move duplication in services to Connection.get and Connection.post

### PR DESCRIPTION
**Pull request by Subskii**
**Motivation (Why)**

Tony and I will soon need to `raise AccessTokenRejected` and limiting the duplication to the `Connection` methods helps a lot with that.

Later we can reduce the duplication in `Connection` too - just not yet.

**Consequence**

Screenshot from 2020-04-30 08-52-14

No change in terms of callers as `Connection` is encapsulated.

### CODE REVIEW

File: lib/xpm_ruby/connection.rb [hunk 1]
```diff
@@ -12,12 +12,48 @@ def initialize(access_token:, xero_tenant_id:)
...
+        case xml["Response"]["Status"]
+        when "OK"
+          xml["Response"]
+        when "ERROR"
+          raise Error.new(response["ErrorDescription"])
+        end
+      else
+        raise Error.new(response.status)
+      end
     end
```
Comment (Subskii) [inline lib/xpm_ruby/connection.rb:57]: As I mentioned in the description we can clean this up later at least it's in once place now.

**Comment by Subskii** [comment]
Think we made a good call resisting the temptation to DRY this early on too @abreckner :smile:.

File: lib/xpm_ruby/client.rb [hunk 2]
```diff
@@ -49,27 +49,9 @@ def build(**args)
...
-        case hash["Response"]["Status"]
-        when "OK"
-          hash["Response"]["Client"]
-        when "ERROR"
-          raise Error.new(response["ErrorDescription"])
-        end
-      else
-        raise Error.new(response.status)
-      end
+        .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))["Client"]
```
Comment (abreckner) [inline lib/xpm_ruby/client.rb:54]: Overall this is great (moving the error handling into the connection which is shared). Much needed! 👍 

Just a small thing, I am just wondering if it's a little more readable to still have a response variable

e.g.

```
response = Connection
        .new(access_token: access_token, xero_tenant_id: xero_tenant_id)
        .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))
response["Client"]
```

I am just wary that if we have too much method chaining the code can get a little hard to follow. We may even want a connection variable too

```
connection = Connection.new(access_token: access_token, xero_tenant_id: xero_tenant_id)
response = connection .post(endpoint: "client.api/add", data: client.to_xml(root: "Client"))
response["Client"]
```
But let me check the Ruby guidelines. This might just be a personal thing for me. ;)

Comment (Subskii) [reply lib/xpm_ruby/client.rb:54]: Like your first example Tony. Seems like a good compromise between too much chaining and too much assignment. I'll update.

Comment (Subskii) [reply lib/xpm_ruby/client.rb:54]: We can remove the chaining on `Connection` later by making it a module instead of an object.

Comment (abreckner) [reply lib/xpm_ruby/client.rb:54]: So I was doing some reading and it appears that method chaining is ok/good practice when you are working on the same object, but as these are returning different objects it can make debugging quite confusing.

e.g. if you were working with Arel where a query is always returned then `Person.where().order().etc()` where it is always working on the query object then it's good practice.

If each method returns a different object it can get confusing.

Elixir has a pipe operator (|>), but Elixir is functional which means you are only doing data transformations with each pipe which is slightly different.

We can have a chat about it in stand up though.
:)

Comment (Subskii) [reply lib/xpm_ruby/client.rb]: Yeap good points I got carried away - have updated :P

**Comment by abreckner** [comment]
Ah I see you updated this. Great stuff!

**Comment by abreckner** [comment]
Reference article: https://stackoverflow.com/questions/1103985/method-chaining-why-is-it-a-good-practice-or-not
